agentprotocol/server.go

```diff
@@ -569,50 +569,19 @@ func (c *ForwardCtx) StartClient() (connectionType uint32, setupPacket SetupPack
 }
 
 func (c *ForwardCtx) StartServerForward() (chan *Connection, error) {
-	c.init()
-
-	setupPacket := SetupPacket{
-		ConnectionType: CONNECTION_TYPE_PORT_DIAL,
-	}
-	mar, err := cbor.Marshal(&setupPacket)
-	if err != nil {
-		c.logger.Error(message.Wrap(
-			err,
-			message.EAgentDecodingFailed,
-			"Error marshalling setup packet",
-		))
-		return nil, err
-	}
-
-	packet := Packet{
-		Type:    PACKET_SETUP,
-		Payload: mar,
-	}
-	err = c.writePacket(&packet)
-	if err != nil {
-		return nil, err
-	}
-
-	resp := Packet{}
-	err = c.decoder.Decode(&resp)
-	if err != nil {
-		return nil, err
-	}
-	if resp.Type == PACKET_ERROR {
-		return nil, fmt.Errorf("received error packet from client")
-	} else if resp.Type != PACKET_SUCCESS {
-		return nil, fmt.Errorf("received invalid packet from client")
-	}
-
-	go c.handleBackend()
-
-	return c.connectionChannel, nil
+	return c.startReverseForwardingClient(nil)
 }
 
-func (c *ForwardCtx) startReverseForwardingClient(setupPacket SetupPacket) (chan *Connection, error) {
+func (c *ForwardCtx) startReverseForwardingClient(setupPacket *SetupPacket) (chan *Connection, error) {
 	c.init()
 
-	mar, err := cbor.Marshal(&setupPacket)
+	if setupPacket == nil {
+		setupPacket = &SetupPacket{
```

Comment:
Why is this packet generated at this function with 'nil' parameter rather than being passed from `StartServerForward` and keeping the pass-by-value? Any specific reason for this?

Reply:
If you pass a value, you need to check every field to see if it is `nil`; but if you pass a pointer, if `nil` check is necessary.

Reply:
> If you pass a value, you need to check every field to see if it is empty; but if you pass a pointer, a nil check is enough.

My point is that the `setupPacket` can be passed down from `StartServerForward` rather than being created here, avoiding the nil check entirely.

Reply:
I see. It seems clearer. I'll fix it later.

Reply:
I see. It seems clearer. I'll fix it later.

Reply:
> Generally, an error return should always be followed by an error check, otherwise we risk loosing the error in subsequent code changes.

Just to be clear, are you misled by the GitHub diff?